drivers/lock.py

```diff
@@ -25,56 +25,39 @@
 class LockException(util.SMException):
     pass
 
-class Lock:
+class Lock(object):
     """Simple file-based lock on a local FS. With shared reader/writer
     attributes."""
 
     BASE_DIR = "/var/lock/sm"
 
-    def _open(self):
-        """Create and open the lockable attribute base, if it doesn't exist.
-        (But don't lock it yet.)"""
-
-        # one directory per namespace
-        self.nspath = os.path.join(Lock.BASE_DIR, self.ns)
-
-        # the lockfile inside that namespace directory per namespace
-        self.lockpath = os.path.join(self.nspath, self.name)
+    INSTANCES = {}
+    BASE_INSTANCES = {}
 
-        number_of_enoent_retries = 10
+    def __new__(cls, name, ns=None, *args, **kwargs):
+        if ns:
+            if ns not in Lock.INSTANCES:
+                Lock.INSTANCES[ns] = {}
+            instances = Lock.INSTANCES[ns]
+        else:
+            instances= Lock.BASE_INSTANCES
 
-        while True:
-            self._mkdirs(self.nspath)
+        if name not in instances:
+            instances[name] = LockImplementation(name, ns)
+        return instances[name]
 
-            try:
-                self._open_lockfile()
-            except IOError, e:
-                # If another lock within the namespace has already
-                # cleaned up the namespace by removing the directory,
-                # _open_lockfile raises an ENOENT, in this case we retry.
-                if e.errno == errno.ENOENT:
-                    if number_of_enoent_retries > 0:
-                        number_of_enoent_retries -= 1
-                        continue
-                raise
-            break
+    # These are required to pacify pylint as it doesn't understancd the __new__
+    def acquire(self):
+        raise NotImplementedError("Lock methods implemented in LockImplementation")
 
-        fd = self.lockfile.fileno()
-        self.lock = flock.WriteLock(fd)
+    def acquireNoblock(self):
```

Comment:
Curious why we need these, does the AttributeError error not contain enough info to debug this?

Reply:
Unless we have an object deriving from this class that i have not got to yet :)

Reply:
See the comment above. pylint claims it to be an error if the method doesn't exist. Because the code calls `lock = Lock('foo')` but actually gets a `LockImplementation` returned to it, which is not and cannot be a subclass of Lock (circular recursion in inheritance if you do that :() Decided this was less ugly than suppressing the pylint error here.